Math/NumberTheory/ArithmeticFunctions/Inverse.hs

```diff
@@ -306,12 +336,40 @@ inverseSigma
   => (a -> b)
   -> a
   -> b
-inverseSigma point = invertFunction point (sigmaA 1) invSigma
+inverseSigma = inverseSigmaK 1
 {-# SPECIALISE inverseSigma :: Semiring b => (Int -> b) -> Int -> b #-}
 {-# SPECIALISE inverseSigma :: Semiring b => (Word -> b) -> Word -> b #-}
 {-# SPECIALISE inverseSigma :: Semiring b => (Integer -> b) -> Integer -> b #-}
 {-# SPECIALISE inverseSigma :: Semiring b => (Natural -> b) -> Natural -> b #-}
 
+-- | The inverse for 'sigma_k' function, where 'k' is a nonnegative integer.
+--
+-- Generalizes the 'inverseSigma' function, which is 'inverseSigmaK' when 'k'
+-- is '1'.
+--
+-- The return value is parameterized by a 'Semiring', which allows
+-- various applications by providing different (multiplicative) embeddings.
+-- E. g., list all preimages (see a helper 'asSetOfPreimages'):
+--
+-- >>> import qualified Data.Set as S
+-- >>> import Data.Semigroup
+-- >>> S.mapMonotonic getProduct (inverseSigmaK 4 (S.singleton . Product) 6651267)
+-- fromList [50]
+--
+-- Similarly to 'inverseSigma', it is possible to count and sum preimages, or
+-- get the maximum/minimum preimage.
+inverseSigmaK
+  :: (Semiring b, Euclidean a, UniqueFactorisation a, Integral a, Enum (Prime a), Bits a)
+  => Word
+  -> (a -> b)
+  -> a
+  -> b
+inverseSigmaK k point = invertFunction point (sigmaA k) invSigma
```

Comment:
You probably need to generalize `invSigma`, making it work for any `k`.

Reply:
`invTotient` worked out of the box for `inverseJordan`, but it seems `invSigma` will need some work.

Could you clarify the generalization that section 5.2 (of the paper you used to write this) had that you removed because `invSigma` only needed the `k = 1` case?

Reply:
`invTotient` works, but mostly by coincidence and doing more work than needed. I think `ps = mapMaybe (isPrime . (+ 1)) divs` should become `ps = mapMaybe (\d -> exactRoot k (d + 1) >>= isPrime) divs`.

With regards to `invSigma` it would probably suffice to replace `p` by `p ^ k` in  conditions in `doPrime` and `pksLarge`.

Reply:
With that in mind, why is [this line](https://github.com/Bodigrim/arithmoi/pull/188/files#diff-54b25104f874d980e04eb1cb0407bc87R154)

`      , e <- [1 .. intToWord (integerLogBase (toInteger lim) (toInteger d))]`

instead of

`      , e <- [1 .. intToWord (integerLogBase (toInteger lim) (toInteger (d - 1)))]`